### ISSUE TITLE
Multiple/redundant queries logged in the console

### ISSUE THREAD

**Issue by achraf-boussaada** [prisma/prisma#7716]
### Bug description

Activating the logs for prisma displays the executed queries multiple times, when the client itself executed the API request only once. Which make me think, this is either a bug, or I'm doing something wrong. here's an example of my logs:
```
prisma:info Fetched a connection from the pool
Query: BEGIN []
Query: BEGIN []
Query: BEGIN []
Query: BEGIN []
Query: BEGIN []
Query: BEGIN []
Query: BEGIN []
Query: BEGIN []
Query: BEGIN []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: DEALLOCATE ALL []
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: SELECT "public"."table"."created_by" FROM "public"."table" WHERE 1=1 GROUP BY "public"."table"."created_by" OFFSET $1 [0]
Query: COMMIT []
Query: COMMIT []
Query: COMMIT []
Query: COMMIT []
Query: COMMIT []
Query: COMMIT []
Query: COMMIT []
Query: COMMIT []
Query: COMMIT []
```

### How to reproduce

I'm using Next.js for the frontend and backend. Defining the client as mentioned below and making a request from the client/insomnia/postman will trigger one API call, which will affect the Database once (I can say that because this happens also with INSERT queries bounded with unique constraints), but the executed queries will be logged multiple times.

### Expected behavior

The query should be logged only once.

### Prisma information

This is how I intiate the prisma client (in a config file globaly and then import it everywhere):
```
import { Prisma, PrismaClient } from '@prisma/client';

const prismaClientOptions: Prisma.PrismaClientOptions = {
  log: [
    {
      emit: 'event',
      level: 'query',
    },
    {
      emit: 'stdout',
      level: 'error',
    },
    {
      emit: 'stdout',
      level: 'info',
    },
    {
      emit: 'stdout',
      level: 'warn',
    },
  ],
};

/* this is needed to avoid creating new client on hot reload of nextjs
 * issue --> https://github.com/prisma/prisma/issues/1983
 */
let prisma: PrismaClient;

if (process.env.NODE_ENV === 'production') {
  prisma = new PrismaClient(prismaClientOptions);
} else {
  //@ts-ignore
  if (!global.prisma) {
    //@ts-ignore
    global.prisma = new PrismaClient(prismaClientOptions);
  }
  //@ts-ignore
  prisma = global.prisma;
}

if (!process.env.NO_PRISMA_LOGGING) { 
  //@ts-ignore
  prisma.$on('query', e => {
    //@ts-ignore
    console.log('Query: ' + e.query, e.params);
    //@ts-ignore
    if (e.duration > 1000) {
      //@ts-ignore
      console.log('Duration: ' + e.duration + 'ms');
    }
  });
}

export { prisma, Prisma };
```
I'm also defining my `DATABASE_URL` with both of the following parameters:
- connection_limit=1
- pgbouncer=true

### Environment & setup

- OS: Debian
- Database: PostgreSQL
- Node.js version: v14.16.0

### Prisma Version

```
prisma               : 2.23.0
@prisma/client       : 2.23.0
Current platform     : debian-openssl-1.1.x
```

**Comment by janpio** [comment]
Is it possible that the `prisma.$on('query', e => {` code you use to prefix and the `console.log` the queries is executed on each request, or multiple times when your app is hot reloaded or similar? It looks to me like your example log you posted above executes that 9 times already. Does it get more over time?

**Comment by achraf-boussaada** [comment]
> It looks to me like your example log you posted above executes that 9 times already. Does it get more over time?

I just noticed that yes, it does get more by time. for this to happen, a new request needs to be made. For instance if I'm on page x where 2 requests will be made they will be then logged twice. If I reload the page they will still be logged twice, if I navigate to other pages where other requests will be executed let's say other 3, then the logs will increase by that number.

> Is it possible that the prisma.$on('query', e => { code you use to prefix and the console.log the queries is executed on each request, or multiple times when your app is hot reloaded or similar?

how can know this? I'm initiating my client in a seperate file and export it as mentioned below and import it in every API route in nextjs.

**Comment by janpio** [comment]
I am not sure how to fix this in your project. But does my explanation make sense to you? Do you think it is possible that the `$on` is executed for each request somehow, hence adding more and more logging logic each time?

**Comment by pantharshit00** [comment]
Can you share actual prisma query calls which are triggering those logs? That will help us debugging this.

**Comment by achraf-boussaada** [comment]
@pantharshit00  here's two examples:

 ```
const x = await prisma.table.groupBy({ by: ['createdBy'] }) 

 SELECT "public"."reports"."created_by" FROM "public"."reports" WHERE 1=1 GROUP BY "public"."reports"."created_by" OFFSET $1 [0]
```
```
 const x = await prisma.$queryRaw`
    SELECT 
      COUNT(id)
    FROM
      table
      ${queryCondition.sql.length > 0 ? Prisma.sql`WHERE ${queryCondition}` : Prisma.empty}
  `
```
the logs of both queries will be repeated the same amount. 

@janpio 

> I am not sure how to fix this in your project. But does my explanation make sense to you? Do you think it is possible that the $on is executed for each request somehow, hence adding more and more logging logic each time?

yes, I think I understand what you mean, but if the client is initiated multiple times or the listener is registered with each query execution. Since I'm doing all of that only once, logically `$on` is only executed once. Or maybe my configuration is somehow faluty here?

**Comment by janpio** [comment]
Yes, that would be my guess. To me it looks like for the `prisma = global.prisma;` case it will add another instance of the logging each time. I think you should move the `$on` call to where you also do `new PrismaClient` to make sure it is only ever executed together.

**Comment by achraf-boussaada** [comment]
I wanted to try something before implementing your suggestion. I just removed `$on` and restarted the server. I noticed this in the logs:
```
prisma:info Fetched a connection from the pool
prisma:info Fetched a connection from the pool
prisma:info Fetched a connection from the pool
prisma:info Fetched a connection from the pool
prisma:info Fetched a connection from the pool
prisma:info Fetched a connection from the pool
prisma:info Fetched a connection from the pool
```
7 entries of prisma fetching a new connection. This is exactly the number of queries executed by the endpoint I called. This seems to me like not a redundant logs, but a more serious problem, where prisma fetch a new connection for every request. Is this normal? I don't explicitly open and close connections in my code, because the [docs](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-management) states that the PrismaClient will manage that for you (in most cases).

Concerning your suggestion I moved the `$on` call to where I do `new PrismaClient`, restarted the server and now I don't get logs anymore.

**Comment by janpio** [comment]
It says that it fetched a connection from the connection pool. That is the only place where it can get a connection from - so this is all correct.

**Comment by achraf-boussaada** [comment]
any other thoughts on this? still couldn't find a working solution

**Comment by janpio** [comment]
I think you optimally open a new issue for the "Fetched from the pool" problem, so we can look at that in isolation. I would consider the redundant query logging problem solved, right? If so, please close this issue.

**Comment by achraf-boussaada** [comment]
Sorry I didn't make myself clear. The `fetched from the pool` problem is cleared, since you mentioned that's how it's supposed to be. I was referring to the original problem, since it's yet not solved. I did as you suggested :

>  I think you should move the $on call to where you also do new PrismaClient to make sure it is only ever executed together.

but still have the same issue.

**Comment by janpio** [comment]
Then the next step would be for you to create a new project, and add all the necessary code to get to a minimal reproduction. From the code snippets you posted I can not identify the problem, so I would need to have a codebase to experience it myself and then look at what might be going on via debugging. That usually is quite effective and lets me skip the "I can not even reproduce this" step when I jsut try to build this myself.
